### ISSUE TITLE
Empty variant

### ISSUE THREAD

**Issue by pschella** [pybind/pybind11#3019]
Would be useful to cast between empty `std::variant` (i.e. `std::monostate`) and `None`. Something like this (in `stl`) should do the trick I think:

```
namespace pybind11 {
namespace detail {
    template <> struct type_caster<std::monostate> {
    public:
        PYBIND11_TYPE_CASTER(std::monostate, _("None"));

        bool load(handle src, bool) {
            return src.ptr() == Py_None;
        }

        static handle cast(std::monostate src, return_value_policy /* policy */, handle /* parent */) {
            Py_RETURN_NONE;
        }
    };
}
}
```

**Comment by SeanCurtis-TRI** [comment]
I completely agree. `std::monostate` is the preferred mechanism for indicating an empty `variant`, it would be nice to have it supported out of the box.

**Comment by Skylion007** [comment]
@SeanCurtis-TRI @pschella Would be happy to review a PR about this with relevant tests.

**Comment by gabrieldevillers** [comment]
Wasn't this fixed by https://github.com/pybind/pybind11/pull/3818 ? @rwgk

**Comment by SeanCurtis-TRI** [comment]
I'm persuaded that the PR closes this issue. You have my vote.

**Comment by rwgk** [comment]
> Wasn't this fixed by #3818 ? @rwgk

Yes, thanks for pointing out!
